src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,469 @@
+#include "deriveNativeScriptHash.h"
+#include "state.h"
+#include "textUtils.h"
+#include "uiScreens.h"
+
+static ins_derive_native_script_hash_context_t* ctx = &(instructionState.deriveNativeScriptHashContext);
+
+// Helper functions
+
+#define TRACE_WITH_CTX(message, ...) TRACE(message "level = %u, remaining scripts = %u", ##__VA_ARGS__, ctx->level, ctx->complexScripts[ctx->level].remainingScripts)
+
+static inline bool areMoreScriptsExpected()
+{
+	// if the number of remaining scripts is not bigger than 0, then this request
+	// is invalid in the current context, as Ledger was not expecting another
+	// script to be parsed
+	return ctx->complexScripts[ctx->level].remainingScripts > 0;
+}
+
+static inline bool isComplexScriptFinished()
+{
+	return ctx->level > 0 && ctx->complexScripts[ctx->level].remainingScripts == 0;
+}
+
+static inline void complexScriptFinished()
+{
+	while (isComplexScriptFinished()) {
+		ASSERT(ctx->level > 0);
+		ctx->level--;
+
+		ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+		ctx->complexScripts[ctx->level].remainingScripts--;
+
+		TRACE_WITH_CTX("complex script finished, ");
+	}
+}
+
+static inline void simpleScriptFinished()
+{
+	ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+	ctx->complexScripts[ctx->level].remainingScripts--;
+
+	TRACE_WITH_CTX("simple script finished, ");
+
+	if (isComplexScriptFinished()) {
+		complexScriptFinished();
+	}
+}
+
+// UI
+
+const char ui_native_script_header[7][27] = {"Script - key path", "Script - key", "Script - ALL", "Script - ANY", "Script - N of K", "Script - invalid before", "Script - invalid hereafter"};
+
+#define ASSERT_UI_SCRIPT_TYPE_SANITY() ASSERT(ctx->ui_scriptType >= UI_SCRIPT_PUBKEY_PATH && ctx->ui_scriptType <= UI_SCRIPT_INVALID_HEREAFTER)
+#define HEADER ui_native_script_header[ctx->ui_scriptType]
+
+static uint8_t _getScriptLevelForPosition()
+{
+	// For complex scripts we reduce the current level by 1
+	// Because they already have the level increased by 1
+	uint8_t levelOffset = ctx->ui_scriptType == UI_SCRIPT_ALL
+	                   || ctx->ui_scriptType == UI_SCRIPT_ANY
+	                   || ctx->ui_scriptType == UI_SCRIPT_N_OF_K
+	                   ? 1 : 0;
+	ASSERT(levelOffset == 0 || ctx->level > 0);
+	return ctx->level - levelOffset;
+}
+
+static void deriveScriptHash_display_ui_position(uint8_t level, ui_callback_fn_t* callback)
+{
+	ASSERT_UI_SCRIPT_TYPE_SANITY();
+	TRACE();
+
+	const size_t POSITION_PREFIX_LENGTH = 11;
+	const size_t MAX_LEVEL_LENGTH = 12;
+
+	// For level n the longest possible position is: "x.x.x...x"
+	// where the x is 2^32-1 (length 10) and there are up to level xs
+	// + the leading "Position: "
+	char positionDescription[10 + 11 * (MAX_SCRIPT_DEPTH - 1) + 1];
```

Comment:
I would prefer to also apply `explicit_bzero` here.

Reply:
Is it any different than using:
```c
char positionDescription[10 + 11 * (MAX_SCRIPT_DEPTH - 1) + 1] = {0};
```
because if I remember correctly this should set all the chars in the array to be zero bytes.

Reply:
It is true according to some standard (C99?). But for some reason, we do not rely on that and use `explicit_bzero` elsewhere. I am not sure how the app is actually compiled and built by Ledger, they might circumvent our compiler options, so I would stick with what `explicit_bzero`.

Reply:
After snprintf, I wouldn't feel safe relying on the standard implementation too much.

Reply:
Added the `explicit_bzero`.